File: 1720. Decode XORed Array.java
/**
 * https://leetcode.com/problems/decode-xored-array/
 */
public class DecodeXORedArray {
    // O(n) time complexity (encoded.length)
    // O(n) space complexity (encoded.length+1)
    int[] decode(int[] encoded, int first) {
        int[] res = new int[encoded.length + 1];
        res[0] = first;

        for (int i = 1; i < res.length; i++) {
            res[i] = res[i - 1] ^ encoded[i - 1];
        }
        return res;
    }
}

class M {
    public static void main(String[] args) {
        int[] encoded = {1, 2, 3};
        int first = 1;
        DecodeXORedArray d = new DecodeXORedArray();
        int[] res = d.decode(encoded, first);
        for (int r : res)
            System.out.print(r + " "); // Ouput: 1 0 2 1
    }
}

File: Kotlin/1720. Decode XORed Array.kt
/**
 * https://leetcode.com/problems/decode-xored-array/
 */
// O(n) time complexity
// O(n) space complexity
fun decode(encoded: IntArray, first: Int): IntArray {
    val res = IntArray(encoded.size + 1)
    res[0] = first

    for (i in 1 until res.size) {
        res[i] = res[i - 1].xor(encoded[i - 1])
    }
    return res
}

fun main() {
    val encoded = intArrayOf(1, 2, 3)
    val first = 1
    val res = decode(encoded, first)
    for (r in res) print("$r ") // Ouput: 1 0 2 1

}